Sources/SwiftSyntaxBuilderGeneration/Templates/BuildableCollectionNodesFile.swift

```diff
@@ -0,0 +1,296 @@
+    signature: FunctionSignature(
+      input: ParameterClause(
+        parameterList: FunctionParameter(
+          firstName: .identifier("arrayLiteral").withTrailingTrivia(.space),
+          secondName: .identifier("elements"),
+          colon: .colon,
+          type: elementType.expressibleAs,
+          ellipsis: .ellipsis
+        )
+      )
+    )
+  ) {
+    FunctionCallExpr(MemberAccessExpr(base: "self", name: "init")) {
+      TupleExprElement(expression: "elements")
+    }
+  }
+}
+
+/// Generate the function building the collection syntax.
+private func createBuildFunction(node: Node) -> FunctionDecl {
+  let type = node.type
+  let elementType = node.collectionElementType
+  return FunctionDecl(
+    modifiers: [TokenSyntax.public],
+    identifier: .identifier("build\(type.baseName)"),
+    signature: FunctionSignature(
+      input: createFormatLeadingTriviaParameters(withDefaultTrivia: true),
+      output: type.syntax
+    )
+  ) {
+    VariableDecl(
+      .let,
+      name: "result",
+      initializer: FunctionCallExpr(MemberAccessExpr(base: "SyntaxFactory", name: "make\(type.baseName)")) {
+        if elementType.isToken {
+          TupleExprElement(expression: "elements")
+        } else {
+          TupleExprElement(
+            expression: FunctionCallExpr(MemberAccessExpr(base: "elements", name: "map"), trailingClosure: ClosureExpr {
+              FunctionCallExpr(MemberAccessExpr(base: "$0", name: "build\(elementType.baseName)")) {
+                TupleExprElement(label: "format", expression: "format")
+                TupleExprElement(
+                  label: "leadingTrivia",
+                  expression: node.elementsSeparatedByNewline
+                    ? SequenceExpr {
+                        MemberAccessExpr(base: "Trivia", name: "newline")
+                        BinaryOperatorExpr("+")
+                        FunctionCallExpr(MemberAccessExpr(base: "format", name: "_makeIndent"))
+                      }
+                    : "nil"
+                )
+              }
+            })
+          )
+        }
+      }
+    )
+    IfStmt(
+      conditions: OptionalBindingCondition(
+        letOrVarKeyword: .let,
+        pattern: "leadingTrivia",
+        initializer: "leadingTrivia"
+      )
+    ) {
+      ReturnStmt(expression: FunctionCallExpr(MemberAccessExpr(base: "result", name: "withLeadingTrivia")) {
+        TupleExprElement(expression: FunctionCallExpr(MemberAccessExpr(
+          base: TupleExpr {
+            SequenceExpr {
+              "leadingTrivia"
+              BinaryOperatorExpr("+")
+              TupleExpr {
+                SequenceExpr {
+                  MemberAccessExpr(base: "result", name: "leadingTrivia")
+                  BinaryOperatorExpr("??")
+                  ArrayExpr(elements: [])
+                }
+              }
+            }
+          },
+          name: "addingSpacingAfterNewlinesIfNeeded"
+        )))
+      })
```

Comment:
Contrary to what I proposed in https://github.com/apple/swift-syntax/pull/506#discussion_r933672865, we may have to stick with the `if-let` for now since the variant I proposed has slightly different semantics (the `if`-version will only consider `result.leadingTrivia` if the function parameter `leadingTrivia` is set). This becomes relevant as otherwise `.addingSpacingAfterNewlinesIfNeeded()` will generate duplicate indentation in some cases.

One alternative solution would be to use optional chaining and only apply the spacing-after-newlines to the function argument `leadingTrivia`:

```swift
let combinedLeadingTrivia = (leadingTrivia?.addingSpacingAfterNewlinesIfNeeded() ?? []) + (result.leadingTrivia ?? [])
return combinedLeadingTrivia
```

instead of the current

```swift
if let leadingTrivia = leadingTrivia {
  return result.withLeadingTrivia((leadingTrivia + (result.leadingTrivia ?? [])).addingSpacingAfterNewlinesIfNeeded())
} else {
  return result
}
```

While that seems to yield the correct result, I am not sure whether there are some edge cases where this logic breaks down (or whether we make too many assumptions about the kind of leading trivia here, e.g. whitespace trivia is 'nice' in the sense that it commutes and therefore may not exercise certain logic errors).

For this reason I decided not to make this change until we have figured that out. Any thoughts on this?

Reply:
> One alternative solution would be to use optional chaining and only apply the spacing-after-newlines to the function argument leadingTrivia

I think that sounds reasonable. I would suggest we discuss it in a separate PR.